content/blog/2021-09-22-refactor.md

```diff
@@ -0,0 +1,162 @@
+﻿---
+title: Easy Stuctural Refactors to Python Source Code
+date: 2021-09-22
+description: |
+  Simple, hassle-free, dependency-free, AST based source code refactoring toolkit.
+
+descriptionLong: |
+  [Refactor](https://github.com/isidentical/refactor) is a source code 
+  refactoring engine. By taking advantage of the Python standard library's
+  [AST](https://docs.python.org/3/library/ast.html) module, we can find-and-fix
+  patterns in massive codebases.
+
+picture: 2021-09-22/ast.png
+author: batuhan_taskaya
+commentsUrl: https://discuss.dvc.org/t/easy-structural-refactors-to-python-source-code/895
+tags:
+  - Refactor
+  - AST
+---
+
+It is common for bigger codebases to grow to a complexity where it is really
+hard for someone to go and refactor stuff manually on every occurence. There are
+automated solutions for this, like regular expression based search and replace
+tools, but the main problem with those is that they treat source code like a
+regular text document, which might lead to some issues like increased number of
+some false positives (a case where the tool did change when it shouldn't), true
+negatives (a case where the tool didn't change when it should have), or even
+not being able to truly express the structural concepts of the programming
+language in them (since many of the programming languages have actual grammars
+which are impossible to manifest in regexes).
+
+This is where [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree)s shine.
+They are the common building blocks of source code; produced by a parser that
+actually understands the language's syntax and creates a tree object where
+smaller parts (e.g. tokens) are ordered in a way that they are related by their
+syntactical meanings.
+
+```python
+password = input("password? ")
+if password == secrets.get("my_password"):
+    print("correct")
+else:
+    print("incorrect")
+```
+
+For example, the AST for the code above will look like this:
+
+![Fundamentals of MLOps](/uploads/images/2021-09-22/ast.png) _Abstract Syntax
+Tree_
+
+The top-most "root" node of this tree represents a single Python file. Each file
+consists of a number of statements (e.g. function definitions, loops, etc.). For
+our example we have only 2 statements: an assignment (to `password`), and an
+`if` statement. Each of these statements in turn have child nodes as defined by
```

Comment:
```suggestion
`if` statement. Each of these statements in turn has child nodes as defined by
```

earlier edited suggestion merged edit :)

Reply:
statements have, not statements has, no?

Reply:
Nope. Each has....  got it.

Reply:
I hated grammar. 😬